src/javascript/utils.js

```diff
@@ -131,3 +132,87 @@ export class UserError extends Error {
     this.title = title;
   }
 }
+
+/**
+ * Given an object containing key-value pairs, produce a URL encodes query
+ * string. This string uses the ?key=value&key2=value2 syntax. If the provided
+ * object is null or empty, return an empty string, otherwise return a string
+ * with a prepended ?
+ */
+export const encodeQuery = query => {
+  if (query) {
+    const encoded = map(query, (value, key) =>
+      `${encodeURIComponent(key)}=${encodeURIComponent(value)}`
+      ).join('&');
+    return encoded ? `?${encoded}` : '';
+  } else {
+    return '';
+  }
+};
+
+// Given a function taking a single argument and returning a promise, wrap
+// that function to memoize the result. The memoize fails if the promise
+// rejects, and multiple concurrent calls to an ongoing promise will all
+// return the same promise.
+export const promiseMemoize = func => {
+  // This is a mapping of key => Promise. storing the promise directly
+  // simplifies our implementation, since we just return it on a cache hit.
+  const cache = new Map();
+
+  return argument => {
```

Comment:
Might not matter, but this only works for functions with 1 argument. This should either be generalized using `arguments...` or added to the documentation.

Reply:
Line 153 says "single argument"? Or did you mean something else?

Reply:
Nope, I just missed that.